### PR TITLE
Improve sheet packing.

### PR DESCRIPTION
**Pull request by RoosterDragon**
When sheet builders are adding sprites to a sheet, they work left to right along each row. They reserve height for the highest sprite seen along that row, resetting the height reservation when the row runs out of space and it moves down to the next row.

As the SpriteCache adds the sprites in a giant batch, it can optimise this operation by ordering the sprites by their height. This reduces wastage where shorter sprites don't use the the full height reserved within the row. The reduced wastage can help the sheet builder allocate fewer sheets, improving load times and improving GPU memory usage as less texture memory is required.

----

By inserting this code after the `SpriteCache.LoadReservations` call in `SequenceSet.LoadSprites` method, you can view the results of this change.

```c#
var count = 0;
foreach (var s in SpriteCache.SheetBuilders[SheetType.BGRA].AllSheets)
	s.AsPng().Save($"{count++}.png");
```

For the Tiberian Dawn HD mod, which uses 8192x8192 sprite sheets, the change reduces the sheet count for the sequence set from 5 to 4.

### CODE REVIEW

File: OpenRA.Game/Graphics/SpriteCache.cs [hunk 1]
```diff
@@ -117,18 +117,11 @@ public void LoadReservations(ModData modData)
...
+							resolvedSprites[token] = resolved;
 							var frames = rs.Frames ?? Enumerable.Range(0, loadedFrames.Length);
 
-							// Premultiplied and non-premultiplied sprites must be cached separately
-							// to cover the case where the same image is requested in both versions.
-							// The premultiplied sprites are stored with an index offset for efficiency
-							// rather than allocating a second dictionary.
-							var di = rs.Premultiplied ? loadedFrames.Length : 0;
 							foreach (var i in frames)
-								resolved[i] = spriteCache.GetOrAdd(i + di,
```
Comment (PunkPun) [inline OpenRA.Game/Graphics/SpriteCache.cs:128]: you are removing the i + di, are you sure you aren't regressing https://github.com/OpenRA/TiberianDawnHD/issues/28 ?

Comment (RoosterDragon) [reply OpenRA.Game/Graphics/SpriteCache.cs:128]: The old version combined the frame index and the premultiplied flag into a single int index. In the new version, we're storing those as distinct elements in the key tuple.

Comment (PunkPun) [reply OpenRA.Game/Graphics/SpriteCache.cs:128]: ah I see

**Review by PunkPun — COMMENTED** [review]
this also reduces TS sheets from 8 -> 6

to test I used

./utility.sh ts --extract isotem.pal
./utility.sh ts --dump-sequence-sheets isotem.pal TEMPERATE

**Review by PunkPun — APPROVED** [review]

**Comment by PunkPun** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bd30de549b3f24dcca7d12308f9e1b2ca7bc8499...9cc927e3c4ad0beb6da030c92c266959d192ecca)

**Pull request merged**